caproto/_commands.py

```diff
@@ -99,60 +99,28 @@ def padded_string_payload(payload):
                      )
 
 
-def data_payload(values, data_type, data_count, *, metadata=None):
-    ntype = native_type(data_type)
-    size = ctypes.sizeof(DBR_TYPES[data_type])
-
-    if ntype != data_type:
-        size += (data_count - 1) * ctypes.sizeof(DBR_TYPES[ntype])
-
-    if data_count != 1:
-        # TODO this needs some work
-        assert data_count == len(values)
-        value_payload = b''.join(map(bytes, values))
+def data_payload(data, metadata, data_type, data_count):
+    if isinstance(data, ctypes.BigEndianStructure):
+        # This is already a DBR.
+        encoded_data = data
+    else:
+        # This is a tuple of values to be encoded into a DBR.
+        ntype = native_type(data_type)
+        cls_dict = {'_fields_': [('value', ntype * data_count)]}
+        cls = type(name, (ctypes.BigEndianStructure,), cls_dict)
```

Comment:
Structures aren't meant for dealing with homogeneously-typed arrays alone.
It's more the job of `struct.pack`:
```python
import struct
struct.pack('>{}i'.format(data_count), *int_array)
```
Or the Python standard `array`, which I've never used somehow: [python docs](https://docs.python.org/3/library/array.html#module-array)
```python
from array import array
arr = array('l', data)
if little_endian:
    arr.byteswap()
```